doc/barclamp.md

```diff
@@ -0,0 +1,1224 @@
+
+Providing these is out of scope for this document. We assume you are able to
+fork the appropriate Barclamp collection. The machines can be provided through
+manual installation or by using an automated setup tool such as
+[mkcloud](https://github.com/SUSE-Cloud/automation/blob/master/scripts/mkcloud).
+
+Optionally (but strongly recommended) you can also
+
+* Create a backup of your Crowbar admin node: You are probably going to break
+  it a few times in the course of testing your new barclamp so it is a good
+  idea to have a backup you can quickly restore rather than rebuild your entire
+  test environment. If you are using `mkcloud` you can create and restore a LVM
+  snapshot of your admin node using its `createadminsnapshot` and
+  `restoreadminfromsnapshot` steps, respectively.
+
+* Create an additional OpenStack controller or compute node as a scratch node:
+  this will allow you to deploy only your own barclamp's controller side
+  components to this node while all other OpenStack services reside on the main
+  controller. This greatly shortens the time a chef run for your barclamp's
+  controller side components takes and thus speeds up development. Make a
+  backup of this node as well to guard against breakage and allow for
+  clean-slate deployment of your barclamp's controller side components.
+
+The best way to set up your scratch node is to make it a lonely node. To this
+end, you will need to re-run mkcloud after you have created your cloud:
+
+```
+nodenumberlonelynode=1 mkcloud setuplonelynodes crowbar_register
+
+```
+
+## Development Environment
+
+Most of your development will take place on your Crowbar node (log in as root).
+You will find the currently running Crowbar code (both Rails application
+components and chef recipes) in the `/opt/dell` directory. We will provide
+detailed locations for all important components of a Barclamp in the _Basic
+Implementation_ section below.
+
+In order to track changes you make we recommend you keep a checkout of your
+`crowbar-openstack` fork on the Crowbar node and apply the changes you make in
+there
+
+### Preparations for Crowbar Node
+
+These are recommendations, but we'll assume you followed them for the remainder
+of this document since this shortens the commands in our examples. First of
+all, clone your `crowbar-openstack` fork to `/root/crowbar-openstack` on your
+Crowbar node and switch to your development branch.
+
+Next, create the following /root/.bashrc with a few useful aliases and
+environment variables (substitute with appropriate values where so indicated by
+comments):
+
+```
+# Replace "Crowbar Committer" by your name as it should appear in commit
+# messages.
+C_NAME="Crowbar Committer"
+
+# Replace "email@address.invalid" by your email address as it should appear in
+# commit messages.
+C_EMAIL="email@address.invalid"
+
+# Replace "6cefbd80254141a63f4c86df195cd930ac7eb10b" by your first commit's ID.
+export COMMIT="6cefbd80254141a63f4c86df195cd930ac7eb10b"
```

Comment:
Yes, it's used to create the diff. One could use something like `git log master..HEAD --oneline | tail -n 1 | cut -f 1`, I suppose, but that relies on the state of the `master` branch being somewhat sensible: if somebody did their changes in their fork's `master` branch or merged part of their topic branch's history to their fork's `master` branch it will blow up. People being people this is likely to happen at some point, and a hardwired commit ID will still Just Work if somebody messes up their local `master` branch.